### PR TITLE
Adding resiliency tests and relevant changes for QueueDAO.

### PR DESCRIPTION
**Pull request by kishorebanala**

### CODE REVIEW

File: core/src/main/java/com/netflix/conductor/core/execution/WorkflowExecutor.java [hunk 1]
```diff
@@ -514,7 +514,23 @@ public void rewind(String workflowId, boolean useLatestDefinitions) {
...
+        try {
+            executionDAOFacade.createWorkflow(workflow);
+        } catch (Exception e) {
+            Monitors.recordWorkflowStartError(workflowDef.getName(), WorkflowContext.get().getClientApp());
+            LOGGER.error("Unable to restart workflow: {}", workflowDef.getName(), e);
+
+            // It's possible the terminate workflow call hits an exception as well, in that case we want to log both
+            // errors to help diagnosis.
+            try {
+                terminateWorkflow(workflowId, "Error when restarting the workflow");
```
Comment (apanicker-nflx) [inline core/src/main/java/com/netflix/conductor/core/execution/WorkflowExecutor.java:527]: I am not convinced that a ```try...catch``` block here is the right way to address this. The exception when terminating a workflow should be caught at a granular level within the terminate logic and this should be logged/handled accordingly.

File: test-harness/src/test/groovy/com/netflix/conductor/test/resiliency/QueueResiliencySpec.groovy [hunk 2]
```diff
@@ -0,0 +1,557 @@
+package com.netflix.conductor.test.resiliency
```
Comment (apanicker-nflx) [inline test-harness/src/test/groovy/com/netflix/conductor/test/resiliency/QueueResiliencySpec.groovy:1]: Please add license header.

File: test-harness/src/test/groovy/com/netflix/conductor/test/resiliency/QueueResiliencySpec.groovy [hunk 3]
```diff
@@ -0,0 +1,557 @@
...
+        def pollResult = taskResource.poll("integration_task_1", "test", "")
+
+        then:
+        1 * queueDAO.pop(*_) >> { throw new IllegalStateException("Queue pop failed from Spy") }
+        0 * queueDAO._
+        notThrown(Exception)
+        pollResult == null
+    }
+
+    def "Verify updateTask succeeds when QueueDAO is unavailable"() {
```
Comment (apanicker-nflx) [inline test-harness/src/test/groovy/com/netflix/conductor/test/resiliency/QueueResiliencySpec.groovy:420]: Do you want to add another test case for updateTask with IN_PROGRESS state? 
This case only covers the updateTask with COMPLETED state, which does not directly use the queue, but the former uses queue to postpone.

File: core/src/main/java/com/netflix/conductor/core/execution/WorkflowExecutor.java (hunk 1 above)
Comment (kishorebanala) [reply core/src/main/java/com/netflix/conductor/core/execution/WorkflowExecutor.java:527]: This exception is primarily for `executionDAOFacade.createWorkflow(workflow)`, and `terminateWorkflow` is just a clean up as part of that exception. I'd still like to throw the exception for `createWorkflow`, irrespective of what happened with `terminateWorkflow`, while logging `terminateWorkflow` exception. Thoughts?

Comment (apanicker-nflx) [reply core/src/main/java/com/netflix/conductor/core/execution/WorkflowExecutor.java:527]: Yes, the exception handling for createWorkflow needs to be handled. 
My comment is regarding the extended handling for the terminate part which seems out of place here. Something like this - 


```java
try {
    executionDAOFacade.createWorkflow(workflow);
} catch (Exception e) {
    Monitors.recordWorkflowStartError(workflowDef.getName(), WorkflowContext.get().getClientApp());
    LOGGER.error("Unable to restart workflow: {}", workflowDef.getName(), e);
    terminateWorkflow(workflowId, "Error when restarting the workflow");
    throw e;
}
```

File: test-harness/src/test/groovy/com/netflix/conductor/test/resiliency/QueueResiliencySpec.groovy [hunk 4]
```diff
@@ -0,0 +1,605 @@
...
+        task.status == Task.Status.IN_PROGRESS
+        task.taskType == 'integration_task_1'
+
+        when: "the above task is updated, while QueueDAO is unavailable"
+        def taskResult = new TaskResult(task)
+        taskResult.setStatus(TaskResult.Status.IN_PROGRESS)
+        taskResult.setCallbackAfterSeconds(120)
+        def result = taskResource.updateTask(taskResult)
+
+        then: "updateTask returns successfully without any exceptions"
```
Comment (apanicker-nflx) [inline test-harness/src/test/groovy/com/netflix/conductor/test/resiliency/QueueResiliencySpec.groovy:496]: update the text here?

File: test-harness/src/test/groovy/com/netflix/conductor/test/resiliency/QueueResiliencySpec.groovy [hunk 5]
```diff
@@ -0,0 +1,605 @@
...
+        2 * queueDAO.postpone(*_) >> { throw new IllegalStateException("Queue postpone failed from Spy") }
```
Comment (apanicker-nflx) [inline test-harness/src/test/groovy/com/netflix/conductor/test/resiliency/QueueResiliencySpec.groovy:497]: Why is postpone invoked twice? Given that we retry twice, shouldn't this be 3?

Comment (kishorebanala) [reply test-harness/src/test/groovy/com/netflix/conductor/test/resiliency/QueueResiliencySpec.groovy:497]: The RetryUtil we're using here retries up to the provided number, which should be greater than >=1.

**Review by apanicker-nflx — APPROVED** [review]

**Pull request merged**